src/moscot/problems/_utils.py

```diff
@@ -52,30 +53,48 @@ def handle_cost(
     x: Optional[Mapping[str, Any]] = None,
     y: Optional[Mapping[str, Any]] = None,
     cost: Optional[Union[CostFn_t, Mapping[str, CostFn_t]]] = None,
+    cost_kwargs: Union[Mapping[str, Any], Mapping[str, Mapping[str, Any]]] = types.MappingProxyType({}),
     **_: Any,
-) -> Tuple[Optional[Mapping[str, Any]], Optional[Mapping[str, Any]], Optional[Mapping[str, Any]]]:
+) -> Tuple[Optional[Dict[str, Any]], Optional[Dict[str, Any]], Optional[Dict[str, Any]]]:
     if cost is None:
-        return xy, x, y
+        return xy, x, y  # type:ignore[return-value]
     if isinstance(cost, str):
         if xy is not None and "cost" not in xy:
             xy = dict(xy)
-            xy["cost"] = cost
-        if x is not None and "cost" not in x:
+            xy["x_cost"] = cost
+            xy["y_cost"] = cost
+        if not (x is None or len(x) == 0) and "cost" not in x:
             x = dict(x)
             x["cost"] = cost
-        if y is not None and "cost" not in y:
+        if not (y is None or len(y) == 0) and "cost" not in y:
             y = dict(y)
             y["cost"] = cost
-        return xy, x, y
-    if isinstance(cost, Mapping):
+    elif isinstance(cost, Mapping):
         if xy is not None and "cost" not in xy:
             xy = dict(xy)
-            xy["cost"] = cost["xy"]
-        if x is not None and "cost" not in x:
+            xy["x_cost"] = cost["xy"]
+            xy["y_cost"] = cost["xy"]
+        if not (x is None or len(x) == 0) and "cost" not in x:
             x = dict(x)
             x["cost"] = cost["x"]
-        if y is not None and "cost" not in y:
+        if not (y is None or len(y) == 0) and "cost" not in y:
             y = dict(y)
             y["cost"] = cost["y"]
-        return xy, x, y
-    raise TypeError(type(cost))
+    else:
+        raise TypeError(type(cost))
+    if xy is not None and len(cost_kwargs):
```

Comment:
This function is getting a bit complicated, please add some comments to explain the behavior (or documentation).

Reply:
Or even better, refactor this into 2 functions, if possible (`handle_linear` and `handle_quadratic`) which are smaller.

Reply:
I tried to simplify a bit and write comments. should be easier now